lib/vrl/stdlib/src/chunks.rs

```diff
@@ -0,0 +1,134 @@
+use vrl::prelude::*;
+
+fn chunks(value: Value, chunk_size: Value) -> Resolved {
+    let bytes = value.try_bytes()?;
+    let chunk_size = chunk_size.try_integer()?;
+
+    if chunk_size < 1 {
+        return Err(r#""chunk_size" must be at least 1 byte"#.into());
+    }
+
+    if let Ok(chunk_size) = usize::try_from(chunk_size) {
+        Ok(bytes.chunks(chunk_size).collect::<Vec<_>>().into())
+    } else {
+        Err(r#""chunk_size" must be a valid usize for this target architecture"#.into())
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct Chunks;
+
+impl Function for Chunks {
+    fn identifier(&self) -> &'static str {
+        "chunks"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::BYTES,
+                required: true,
+            },
+            Parameter {
+                keyword: "chunk_size",
+                kind: kind::INTEGER,
+                required: true,
+            },
+        ]
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "chunks by byte",
+                source: r#"chunks("abcdefgh", 4)"#,
+                result: Ok(r#"["abcd", "efgh"]"#),
+            },
+            Example {
+                title: "chunk size must be at least 1 byte",
+                source: r#"chunks("hello world", 0)"#,
+                result: Err(
+                    r#"function call error for "chunks" at (0:24): "chunk_size" must be at least 1 byte"#,
+                ),
+            },
+            Example {
+                title: "chunk sizes do not respect unicode code point boundaries",
+                source: r#"chunks("ab你好", 4)"#,
+                result: Ok(r#"["ab�","�好"]"#),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let chunk_size = arguments.required("chunk_size");
+
+        Ok(Box::new(ChunksFn { value, chunk_size }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ChunksFn {
+    value: Box<dyn Expression>,
+    chunk_size: Box<dyn Expression>,
+}
+
+impl Expression for ChunksFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let chunk_size = self.chunk_size.resolve(ctx)?;
+
+        chunks(value, chunk_size)
+    }
+
+    fn type_def(&self, _state: (&state::LocalEnv, &state::ExternalEnv)) -> TypeDef {
+        let chunk_size_is_expression = self.chunk_size.as_value().is_none();
+        // chunk_size is converted to a usize later, so if a user-supplied Value::Integer (i64) is
+        // larger than the platform's usize::MAX, it could fail to convert.
+        let chunk_size_limited_by_architecture = usize::BITS < 64;
```

Comment:
Currently if a _valid_ literal is passed in on a 32-bit system, the function will be marked infallible. The `usize::BITS` value should probably only be used when checking if a literal is in the valid range. If you don't have a literal, the function should already be marked as `infallible` (since we don't know what the input could be)

Reply:
> If you don't have a literal, the function should already be marked as `infallible` (since we don't know what the input could be)

I may be misunderstanding what `infallible` means, but do you mean `fallible` here?

Reply:
> I may be misunderstanding what infallible means, but do you mean fallible here?

Yep, sorry! I fixed it.

Reply:
So if the user supplies a literal `chunk_size` that we know is not within the valid range, can we say definitively at that point that `chunks` will fail? As a user, I think I would prefer it to be a compile time error rather than handling the fallibility and then also discovering the error at runtime, especially when it's something like 0 or \*checks notes\* 4294967296 (~4GiB).

As a side note, I think using `usize::BITS` may be a bit of a red herring since the easiest way to see if the chunk size is within a valid range is to use `usize::try_from(chunk_size)` and `chunk_size > 0`. `usize::BITS` tells us about the platform's `usize`, but we can just check if it's a valid `usize` directly.

My proposal:
1. Make it fallible if it's an expression
2. Make it a compile time error if it's an out-of-range literal
3. Else make it infallible

I think the compile time checks would be done in `fn compile`, right? However from our call, I got the impression that that compile time checks on the literal values aren't generally done that way? Let me know what you think.

Reply:
> So if the user supplies a literal chunk_size that we know is not within the valid range, can we say definitively at that point that chunks will fail? As a user, I think I would prefer it to be a compile time error rather than handling the fallibility and then also discovering the error at runtime

That's reasonable. I'm fine with failing with a compile time error if it's guaranteed to fail.

> As a side note, I think using usize::BITS may be a bit of a red herring since the easiest way to see if the chunk size is within a valid range is to use usize::try_from(chunk_size) and chunk_size > 0. usize::BITS tells us about the platform's usize, but we can just check if it's a valid usize directly.

Right. Using `usize::try_from` should be sufficient.

Overall your proposal is reasonable, although I would re-word 1 to "Make it fallible if it's not a literal" (since _all_ arguments are expressions)

> I think the compile time checks would be done in fn compile, right? However from our call, I got the impression that that compile time checks on the literal values aren't generally done that way? Let me know what you think.

Correct. We don't do this today in many places (for example `1/0` compiles, but is fallible), but I'm not opposed to adding it. I'm considering expanding the "fallibility" concept in the future to be an enum of something like `infallible` / `fallible` / `always_fails`, and then the compiler can automatically error if it's guaranteed to fail, but for now adding it in the `compile` function is fine.

Reply:
> I'm considering expanding the "fallibility" concept in the future to be an enum of something like `infallible` / `fallible` / `always_fails`, and then the compiler can automatically error if it's guaranteed to fail

I think this would be helpful in this case, because then I could check whether `chunk_size` is in range in one place instead of 3.